### PR TITLE
Continue localstack test harness development

### PR DESCRIPTION
**Pull request by sjperkins**
Follows on from these comments in #91

- https://github.com/google/tensorstore/pull/91#pullrequestreview-1555903915
- https://github.com/google/tensorstore/pull/91#pullrequestreview-1555898558
- https://github.com/google/tensorstore/pull/91#pullrequestreview-1554191686

### CODE REVIEW

**Comment by laramiel** [comment]
Heh. I just submitted a few updates which conflict with this--it separates out the localstack tests from the other tests, so your changes will need to target the new file.

**Comment by sjperkins** [comment]
Setting `options.byte_range.exclusive_max = 2;` here:

https://github.com/google/tensorstore/blob/78b93c9c0f43a0e38e00bcdfee3a7783d1c833ad/tensorstore/kvstore/test_util.cc#L442-L450

to produce a `Range: bytes=1-1` produces the 206 but still fails the assert:



```
I0000 00:00:1693902191.704033      81 s3_key_value_store.cc:515] ReadTask: HttpRequest{GET http://localhost:4566/tensorstore/test/key_read user_agent=, Range: bytes=1-1, host: testbucket.s3.af-south-1.localstack.localhost.com, x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855, x-amz-date: 20230905T082311Z, Authorization: AWS4-HMAC-SHA256 Credential=LSIAQAAAAAAVNCBMPNSG/20230905/af-south-1/s3/aws4_request,SignedHeaders=host;range;x-amz-content-sha256;x-amz-date,Signature=535c37221d7c17ea32453c98a87f896bef2e0518f55f141c312df08f8d7ad31a}
I0000 00:00:1693902191.763336      80 s3_key_value_store.cc:528] ReadTask HttpResponse{code=206, headers=<accept-ranges=bytes, content-range=bytes 1-1/15, content-type=application/octet-stream, date=Tue, 05 Sep 2023 08:23:11 GMT, etag="6f314ce98d753057568807b0863b3bf0", last-modified=Tue, 05 Sep 2023 08:23:11 GMT, server=hypercorn-h2, x-amz-id-2=s9lzHYrFp76ZVxRcpX9+5cjAnEH2ROuNkd2BHfIa6UkFVdtjf5mKR3/eTPFvsiP/XV/VLi31234=, x-amz-request-id=33042dbb-1e20-44e1-ae8b-d2f7092d2482>, body=k}
tensorstore/kvstore/test_util.cc:447: Failure
Value of: kvstore::Read(store, key, options).result()
Expected: value (is an object whose field `state` is equal to <value>) and (is an object whose field `value` is equal to ) and (is an object whose field `stamp` value (is an object whose field `generation` is equal to "\"6f314ce98d753057568807b0863b3bf0\"\x01") and (is an object whose field `time` is anything))
  Actual: Result{
{value="k", stamp={generation="\"6f314ce98d753057568807b0863b3bf0\"\x01", time=2023-09-05T08:23:11.703827803+00:00}}},
whose value
{value="k", stamp={generation="\"6f314ce98d753057568807b0863b3bf0\"\x01", time=2023-09-05T08:23:11.703827803+00:00}} doesn't match,
whose field `value` is k (of type absl::Cord)
Google Test trace:
tensorstore/kvstore/test_util.cc:400: TestKeyValueStoreReadOps
Stack trace:
  0x7f9e3bbc1d34: tensorstore::internal::TestKeyValueStoreReadOps()
  0x7f9e3bbc5d82: tensorstore::internal::TestKeyValueReadWriteOps()
  0x7f9e3bbc5933: tensorstore::internal::TestKeyValueReadWriteOps()
  0x563c60abcf8f: (anonymous namespace)::LocalStackFixture_Basic_Test::TestBody()
  0x7f9e3a98dfd1: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x7f9e3a987e7b: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x7f9e3a969ad4: testing::Test::Run()
  0x7f9e3a96a562: testing::TestInfo::Run()
... Google Test internal frames ...
```

File: tensorstore/kvstore/s3/s3_key_value_store_test.cc [hunk 1]
```diff
@@ -59,15 +59,15 @@ using ::tensorstore::internal_http::HttpResponse;
 using ::tensorstore::internal_kvstore_s3::S3Credentials;
 using ::tensorstore::internal_kvstore_s3::S3RequestBuilder;
 
-ABSL_FLAG(std::string, localstack_binary, "", "Path to the localstack");
+ABSL_FLAG(std::string, localstack_binary, "", "Path to the localstack binary");
+ABSL_FLAG(std::string, localstack_endpoint, "http://localhost:4566", "Localstack endpoint");
```
[the remark this answers is not in the document]
Comment (sjperkins) [reply tensorstore/kvstore/s3/s3_key_value_store_test.cc:63]: Looks like its configurable via environment variables 

- https://docs.localstack.cloud/references/configuration/
- `GATEWAY_LISTEN`
- `LOCALSTACK_HOST`

I guess its possible to try `PickUnusedPortOrDie` as discussed here: https://github.com/google/tensorstore/pull/91#discussion_r1279950946

**Comment by sjperkins** [comment]
> Heh. I just submitted a few updates which conflict with this--it separates out the localstack tests from the other tests, so your changes will need to target the new file.

:-) We should probably coordinate more -- apologies I was out on PTO mid-August. Thanks for merging #91 and adding the docs.

I can devote a day or so a week to working on tensorstore so the velocity is not as high as it could be. I'd imagine the context switches aren't ideal for either of us. If it's desirable to develop the remaining S3 functionality more speedily, I'd be fine with stepping back and let others take the helm. Alternatively, I can continue at the current pace, aiming to get the localstack testharness into good shape. Then, the remaining work should be easier for others to deal with piecemeal in smaller PRs.

Also, In terms of communication, I think it would be valuable if I converted the TODO in #91 into an umbrella issue detailing the remaining S3 work.

File: tensorstore/kvstore/s3/validate_test.cc [hunk 2]
```diff
@@ -33,14 +33,17 @@ TEST(ValidateTest, ClassifyBucketName) {
   EXPECT_EQ(ClassifyBucketName("sthree-configurator.bucket"),
             BucketNameType::kInvalid);
 
-  // Standard bucket names are < 63 characters
+    // Standard bucket names are < 63 characters
```
Comment (laramiel) [inline tensorstore/kvstore/s3/validate_test.cc:36]: Split this test change into a separate pull request, please.

File: tensorstore/kvstore/s3/s3_key_value_store_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (laramiel) [reply tensorstore/kvstore/s3/s3_key_value_store_test.cc:63]: Yes, if it's unset (and the env var is unset), that would be nice.

File: tensorstore/kvstore/s3/BUILD [hunk 3]
```diff
@@ -173,12 +173,16 @@ tensorstore_cc_test(
     name = "s3_key_value_store_test",
     size = "small",
     srcs = ["s3_key_value_store_test.cc"],
+    # args = ["--localstack_binary=$(location @pypa_localstack//:localstack)"],
```
[the remark this answers is not in the document]
Comment (laramiel) [reply tensorstore/kvstore/s3/BUILD:176]: To do that you'd need a workspace for localstack with a py_binary, similar to what we have for storagebench:

https://github.com/google/tensorstore/blob/master/third_party/com_google_storagetestbench/workspace.bzl

Then reference it like here, similar to what the gcs_grpc_testbench_test:

https://github.com/google/tensorstore/blob/97bd317d182df8789aba98d7b0421458e706c791/tensorstore/kvstore/gcs_grpc/BUILD#L129

I'd call the directory com_github_localstack or perhaps cloud_localstack.

**Comment by laramiel** [comment]
There are a few cleanups that I was going to do, mostly related to using Context for AWS auth, and perhaps a very minimal mock based test (similar to the gcs driver). But since it's submitted, the s3 kvstore will also be included in our continuous build and will see updates that affect it in that way, which might include updates which affect all of our drivers.

But other than that, I don't have immediate plans to add features and am happy let you handle those improvements. Hopefully we won't step on each other's toes too much.

**Comment by laramiel** [comment]
w.r.t the range issue, I had to fix a few of those in the gcs testbench, it seems very likely that range requests are not well tested in localstack as well. If AWS respond correctly, it seems like a pull request for localstack is worthwhile.

**Comment by laramiel** [comment]
Just a note about localstack_test vs other possibilities.

With the addition of `"s3"` to [tensorstore/kvstore/BUILD](https://github.com/google/tensorstore/blob/5eee05605af8dd380d40611e0c7c2e4f1d5795dc/tensorstore/kvstore/BUILD#L8)

Then the following will work to access an actual s3 bucket:

```
./bazelisk.py build --//tensorstore/kvstore/s3:debug //tensorstore/kvstore:live_kvstore_test

AWS_ACCESS_KEY_ID=... AWS_SECRET_ACCESS_KEY=... \
bazel-bin/tensorstore/kvstore/live_kvstore_test --kvstore_spec='{"driver": "s3", "bucket": "MYBUCKET"}'
```

**Comment by sjperkins** [comment]
> Briefly, it looks like AWS behaviourly treats 1-0 as a request for the entire content.

How does GCS respond to this?

**Comment by sjperkins** [comment]
> For localstack, the most important thing is that it mimics AWS responses, and AWS responds like this:

> Which is the entire value.

The localstack devs have indicated on the issue they agree that it should return the entire response.

**Comment by laramiel** [comment]
I have a pending update for http-style backends to avoid sending invalid Range: headers.

**Comment by sjperkins** [comment]
I had a closer look at the [rfc](https://www.rfc-editor.org/rfc/rfc9110.html#name-range-requests) and it looks like a server may choose to ignore an invalid range request:

> A server that supports range requests MAY ignore or reject a [Range](https://www.rfc-editor.org/rfc/rfc9110.html#field.range) header field that contains an invalid [ranges-specifier](https://www.rfc-editor.org/rfc/rfc9110.html#rule.ranges-specifier) ([Section 14.1.1](https://www.rfc-editor.org/rfc/rfc9110.html#range.specifiers)), a [ranges-specifier](https://www.rfc-editor.org/rfc/rfc9110.html#rule.ranges-specifier) with more than two overlapping ranges, or a set of many small ranges that are not listed in ascending order,

**Comment by sjperkins** [comment]
The AWS range issue fix was merged:

- https://github.com/localstack/localstack/issues/9076
- https://github.com/localstack/localstack/pull/9082

I guess it will be included in the 2.3 release, which does not seem to have many remaining issues:

- https://github.com/localstack/localstack/pulls?q=is%3Apr+milestone%3A2.3+is%3Aopen

I'll likely recreate this PR once 2.3 is released.

**Comment by sjperkins** [comment]
FWIW I've run the localstack test case locally with the `localstack/localstack:latest` image (c.f. https://github.com/localstack/localstack/issues/9076#issuecomment-1711974643) and `tensorstore::internal::TestKeyValueReadWriteOps` succeeds.

Looks like localstack 2.3 is targeting a 26th September release:

- https://github.com/localstack/localstack/milestone/4

**Comment by laramiel** [comment]
You should be able to use even the older version now since I've updated the range request logic across the board to avoid the bad requests.

**Comment by laramiel** [comment]
With my latest commit it's possible to do this (on a mac with python 3.10):

```
python3 -m pip install localstack[runtime]
./bazelisk.py run //tensorstore/kvstore/s3:localstack_test -- \
   --localstack_binary=$(which localstack)
```
